Fatras/include/ActsFatras/Digitization/UncorrelatedHitSmearer.hpp

```diff
@@ -0,0 +1,182 @@
+  /// @param volume The Volume in which this measurement was measured
+  /// @param sFunctions The smearing functions that are applied
+  ///
+  /// @return  A surface based Measurement, or a null result
+  template <typename hit_t, Acts::FreeIndices... params>
+  Acts::Result<VolumeMeasurement<hit_t, params...>> createVolumeMeasurement(
+      const hit_t& hit, std::shared_ptr<const Acts::Volume> volume,
+      const std::array<SmearFunction, sizeof...(params)>& sFunctions) const {
+    using Result = Acts::Result<VolumeMeasurement<hit_t, params...>>;
+
+    auto pos = hit.position();
+    auto dir = hit.unitDirection();
+
+    Acts::FreeVector freeVector;
+    freeVector[Acts::eFreePos0] = pos.x();
+    freeVector[Acts::eFreePos1] = pos.y();
+    freeVector[Acts::eFreePos2] = pos.z();
+    freeVector[Acts::eFreeTime] = hit.time();
+    freeVector[Acts::eFreeDir0] = dir.x();
+    freeVector[Acts::eFreeDir1] = dir.y();
+    freeVector[Acts::eFreeDir2] = dir.z();
+
+    typename VolumeMeasurement<hit_t, params...>::ParameterVector vector;
+    vector.setZero();
+
+    typename VolumeMeasurement<hit_t, params...>::CovarianceMatrix covariance;
+    covariance.setZero();
+
+    auto sResult = smear(freeVector, vector, covariance, sFunctions, params...);
+    if (not sResult.ok()) {
+      return Result(sResult.error());
+    }
+
+    return Result(VolumeMeasurement<hit_t, params...>(
+        volume, hit, std::move(covariance), vector));
+  }
+
+ private:
+  /// Apply the recursive smearing on the parameter pack,
+  /// the base local position or time is used as unsmeared
+  ///
+  /// @tparam full_vec_t The full parameter vector, unsmeared
+  /// @tparam vec_t The parameter vector, static size known at compile time
+  /// @tparam cov_t The covariance materix, static size known at compile time
+  /// @tparam smear_t The smearing funtion array, static size known at c.t.
+  /// @tparam param_t The first of the pack
+  /// @tparam paramts_t The remaining pack
+  ///
+  /// @param fVector The original unsmeared hit information (full dimensional)
+  /// @param vector The to be filled parameters vector
+  /// @param covariance The to be filled covariance matrix
+  /// @param smearing The array of smearing functions
+  /// @param first The first of the pack
+  /// @param rest The remaining pack
+  ///
+  /// @return boolean to indiciate it was successful
+  template <typename full_vec_t, typename vec_t, typename cov_t,
+            typename smear_t, typename param_t, typename... params_t>
+  Acts::Result<void> smear(const full_vec_t& fVector, vec_t& vector,
+                           cov_t& covariance, smear_t& smearing,
+                           const param_t& first,
+                           params_t const&... rest) const {
+    const size_t par = vector.rows() - sizeof...(rest) - 1;
+    auto sResult = smearing[par]();
+    if (sResult.ok()) {
+      const auto& sr = sResult.value();
+      vector[par] = fVector[first] + sr.first;
```

Comment:
Should we do the `SmearOutOfBounds' check here? Or it's supposed to be handled by the `smearing` function somehow.

Reply:
Indeed, the trick is that the `smearing[par]` function can return 'SmearOutOfBounds`, I plan to have the following smearing struct:
- Gaussian (bound & clipped)
- Gaussian (unbound, e.g. for angular smearing(
- MultiGaussian (in above flavours)

So, it's up to the smearer to return `ok()` or not, and then the error code if needed is forwarded.

Reply:
I see. Thanks!